File: 0685-Redundant Connection II/0685-Redundant Connection II.cpp
class Solution {
public:
    vector<int> findRedundantDirectedConnection(vector<vector<int>>& edges) {
        int N = edges.size();
        vector<int> parent(1 + N);
        vector<int> candA, candB;
        for (auto& edge : edges) {
            if (parent[edge[1]] == 0) {
                parent[edge[1]] = edge[0];
            }
            else {
                candA = {parent[edge[1]], edge[1]};
                candB = edge;
                edge[1] = 0;
                break;
            }
        }
        
        for (int i = 1; i <= N; ++i) {
            parent[i] = i;
        }
        
        for (auto& edge : edges) {
            if (edge[1] == 0) {
                continue;
            }
            
            int u = edge[0];
            int pu = findParent(parent, u);
            int v = edge[1];
            if (pu == v) {
                if (candA.size()) {
                    return candA;
                }
                else {
                    return edge;
                }
            }
            parent[v] = pu;
        }
        
        return candB;
    }
    
private:
    int findParent(vector<int>& parent, int i) {
        while (parent[i] != i) {
            parent[i] = parent[parent[i]];
            i = parent[i];
        }
        
        return i;
    }
};

File: 0601-0700/0685-Redundant Connection II/0685-Redundant Connection II.py
class Solution:
    def findRedundantDirectedConnection(self, edges: List[List[int]]) -> List[int]:
        parent = [i for i in range(len(edges) + 1)]
        def findParent(x):
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x
        
        visited = [-1] * (len(edges) + 1)
        first = second = last = -1
        for i, edge in enumerate(edges):
            n1 = edge[0]
            n2 = edge[1]
            if visited[n2] != -1:
                first = visited[n2]
                second = i
                continue
            visited[n2] = i
            
            p1 = findParent(n1)
            if p1 == n2:
                last = i
            else:
                parent[n2] = p1
                
        if last == -1:
            return edges[second]
        if second == -1:
            return edges[last]
        return edges[first]
